framework/include/utils/SymmetricRankFourTensor.h

```diff
@@ -0,0 +1,581 @@
+ */
+template <typename T>
+class SymmetricRankFourTensorTempl
+{
+public:
+  /// Dimensionality of rank-four tensor
+  static constexpr unsigned int Ndim = 3;
+  static constexpr unsigned int N = 2 * Ndim;
+  static constexpr unsigned int N2 = N * N;
+
+  /// returns the 1, sqrt(2), or 2 prefactor in the Mandel notation
+  static constexpr Real mandelFactor(unsigned int i, unsigned int j)
+  {
+    return i < 3 ? (j < 3 ? 1.0 : MathUtils::sqrt2) : (j < 3 ? MathUtils::sqrt2 : 2.0);
+  }
+
+  /// Initialization method
+  enum InitMethod
+  {
+    initNone,
+    initIdentity,
+    initIdentitySymmetricFour
+  };
+
+  /**
+   * To fill up the 36 entries in the 4th-order tensor, fillFromInputVector
+   * is called with one of the following fill_methods.
+   * See the fill*FromInputVector functions for more details
+   */
+  enum FillMethod
+  {
+    symmetric9,               // fillSymmetric9FromInputVector
+    symmetric21,              // fillSymmetric21FromInputVector
+    symmetric_isotropic,      // fillSymmetricIsotropicFromInputVector
+    symmetric_isotropic_E_nu, // fillSymmetricIsotropicEandNu
+    axisymmetric_rz,          // fillAxisymmetricRZFromInputVector
+    principal,                // fillPrincipalFromInputVector
+    orthotropic               // fillGeneralOrthotropicFromInputVector
+  };
+
+  template <template <typename> class Tensor, typename Scalar>
+  struct TwoTensorMultTraits
+  {
+    static const bool value = false;
+  };
+  template <typename Scalar>
+  struct TwoTensorMultTraits<SymmetricRankTwoTensorTempl, Scalar>
+  {
+    static const bool value = ScalarTraits<Scalar>::value;
+  };
+  template <typename Scalar>
+  struct TwoTensorMultTraits<TensorValue, Scalar>
+  {
+    static const bool value = ScalarTraits<Scalar>::value;
+  };
+  template <typename Scalar>
+  struct TwoTensorMultTraits<TypeTensor, Scalar>
+  {
+    static const bool value = ScalarTraits<Scalar>::value;
+  };
+
+  /// Default constructor; fills to zero
+  SymmetricRankFourTensorTempl();
+
+  /// Select specific initialization pattern
+  SymmetricRankFourTensorTempl(const InitMethod);
+
+  /// Fill from vector
+  SymmetricRankFourTensorTempl(const std::vector<T> &, FillMethod);
+
+  /// Copy assignment operator must be defined if used
+  SymmetricRankFourTensorTempl(const SymmetricRankFourTensorTempl<T> & a) = default;
+
+  /**
+   * Copy constructor
+   */
+  template <typename T2>
+  SymmetricRankFourTensorTempl(const SymmetricRankFourTensorTempl<T2> & copy);
+
+  // explicit cast to a full tensor
+  explicit operator RankFourTensorTempl<T>();
+
+  // Named constructors
+  static SymmetricRankFourTensorTempl<T> Identity()
+  {
+    return SymmetricRankFourTensorTempl<T>(initIdentity);
+  }
+  static SymmetricRankFourTensorTempl<T> IdentitySymmetricFour()
+  {
+    return SymmetricRankFourTensorTempl<T>(initIdentitySymmetricFour);
+  };
+
+  /// Gets the value for the index specified.  Takes index = 0,1,2,3,4,5
```

Comment:
'Gets the value for the indices specified. Takes indices ranging from 0-5 for i and j'?